### ISSUE TITLE
Working DL4J AutoEncoder

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#635]
Since we might need an AutoEncoder for #615 , you should try learning about how to train an AutoEncoder in DL4J, a deep learning platform for Java that is already integrated into the MM-NEAT code. Make a band new package called `autoencoder` and a class called `AutoEncoderExample` and try walking through the tutorial here:

https://deeplearning4j.konduit.ai/getting-started/tutorials/basic-autoencoder

I think that all the required libraries are already part of MM-NEAT, but let me know if you run into problems.

**Comment by schrum2** [comment]
The eventual goal of this is to apply an autoencoder to Picbreeder images. Fortunately, the run of PictureTargetTask that you are doing ( #615 ) is creating lots of images. So, at some point, your goal should be to train an autoencoder using those images to see how well it can reconstruct them.

**Comment by annazwicker** [comment]
My understanding so far: the autoencoder takes an input and compresses it, then tries to rebuild the original input from the compressed version.  The distance function ("loss function") helps to fix errors made by the decoder when recreating the input (could you say the program is learning?).  The autoencoder does this on its own without user assistance.

@schrum2 let me know if this is even remotely correct

**Comment by schrum2** [comment]
The loss function measures the difference between what the image is and what it is supposed to be. In fact, it is very similar to the fitness function we are trying to calculate for the Picbreeder Task. However, there are numerous difference loss functions. However, it occurs to me that if the DL4J code can easily calculate a loss between actual and target outputs, then maybe we could use that as our fitness for the PictureTargetTask #615 . I'll follow up there.

As for autoencoders, once you have a measure of the loss, the actual procedure that trains the network is called backpropagation of error.

**Comment by schrum2** [comment]
Since the basic MNIST example works, I want you to now attempt training an autoencoder on Picbreeder images. You will need to research the internet how to create your own training set out of your own images. You can use images already produced by the preliminary runs of MAP Elites #615 . You don't have to use all of them ... for your prototype, try a smaller training set ... maybe 25 images? I don't really know what an appropriate amount is, but something matching the size of the training set for the MNIST example. I think DL4J also provides a way to downscale images, and you should downscale the picbreeder images to the size of the MNIST example digits, so that you can easily use the same network architecture.

**Comment by schrum2** [comment]
In the class you've named `PrototypePictureTargetTaskAutoEncoder` (by the way, this should probably be given its original name of `ImagePipelineExample`), there is a class `ImageRecordReader` that is responsible for processing all of the loaded images. This is later used to initialize a `DataSetIterator`.

I think that instead of the `RecordReader` that is inside of `PictureTargetTaskAutoEncoder`, you might be able to use this `ImageRecordReader` instead. This should also allow you to avoid making a bunch of extra classes modelled on the MNIST classes.

The basic idea is that you want to do everything the autoencoder code does in the exact same way, except you want to change the source of the data. The data is processed in the MNIST example by the `MnistDataSetIterator`, which is a type of `DataSetIterator`. The code in `PrototypePictureTargetTaskAutoEncoder` also creates a `DataSetIterator`, but for a different data source. So, it you simply take the code for the `DataSetIterator` from `PrototypePictureTargetTaskAutoEncoder` and replace the `MnistDataSetIterator` then maybe it will work.

**Comment by annazwicker** [comment]
Epochs:

- 1: produces noise
- 2: produces solid black images
- 3: produces black images with some gray specks
- 4: produces black images with more gray specks than 3 epochs
- 9: produces differing images, two darker images with medium and light gray specks, one medium gray with dark gray specks, and one light gray with dark gray specks
- 10: produces light gray images with dark specks
- 30, 50, 100, 1000, 5000: all produce relatively similar images (light gray with dark specks).

**Comment by schrum2** [comment]
I'm curious to know what the logged "score" looks like. When the 100,000 epoch run finishes, paste a picture of the images and say what the final logged score was.

Either way, I think the next step in playing around with this should be to see how well the autoencoder recreates the training images. Current, it trains on the training data, and then checks how well it does on each of the test images before displaying the test images. We should also look at how well if does recreating the original training images. To see how the autoencoder performs on the training images, I think that you can just replace the array for the test data with the array for the training data in the section of the code that creates pairs of autoencoder input/outputs.

**Comment by annazwicker** [comment]
Score decreased but image is pretty much the same as 10+ epochs.
Score at iteration 199990 is 195.34823602024113
![5-25-2021 100000 Epochs](https://user-images.githubusercontent.com/82399865/119563829-ac9d3180-bd6d-11eb-9bfd-b092e2c5bc42.PNG)

**Comment by annazwicker** [comment]
10,000 Epochs
Score at iteration 19990 is 612.920857461024
![5-25-2021 10000 Epochs](https://user-images.githubusercontent.com/82399865/119564621-9b085980-bd6e-11eb-95a4-62af9f888b79.PNG)

**Comment by annazwicker** [comment]
50,000
Score at iteration 99990 is 354.46647813589556
![5-25-2021 50000 Epochs](https://user-images.githubusercontent.com/82399865/119565956-33eba480-bd70-11eb-96d7-a34b029ec8fa.PNG)

**Comment by annazwicker** [comment]
Loaded with MNIST Images but received the same result: loads the saved digit images (but not clearly) and cannot reconstruct the images the same way AutoEncoderExample can.
![MNIST Data 1](https://user-images.githubusercontent.com/82399865/119573358-be84d180-bd79-11eb-9f12-f80a23fdc176.PNG)
![MNIST Data 2](https://user-images.githubusercontent.com/82399865/119573359-be84d180-bd79-11eb-8412-d6b04570c51f.PNG)
![MNIST Data 3](https://user-images.githubusercontent.com/82399865/119573360-bf1d6800-bd79-11eb-8aa7-f254338b9e01.PNG)

**Comment by schrum2** [comment]
Although we're stuck on this, I would still like to solve this issue eventually. There is always a chance it could be useful to do things in Java instead of Python ( #655 ), but since we're making better progress with Python at the moment we'll focus there first.

**Comment by schrum2** [comment]
@annazwicker Although we are mostly happy with the Python version of this, I want you to return to trying to get the DL4J Java AutoEncoder to work. If we can get both working, we can compare them. This one could eventually be better if we can get it to work.

Also, now we have lots of example directories full of CPPN images (from these runs #642) that we could train on to troubleshoot things.

**Comment by schrum2** [comment]
Getting the Java autoencoder to work is a good issue to keep in here for the long term, but no need to worry about completing it this summer.

**Comment by schrum2** [comment]
Closing this issue after deleting the code for the Java Autoencoder. The Python autoencoder for the PictureTargetTask still works ... at least, as much as it ever did.
